File: AndroidArchitecture/mvvm-live/src/main/java/com/codearms/maoqiqi/app/addedittask/AddEditTaskActivity.java
package com.codearms.maoqiqi.app.addedittask;

import androidx.lifecycle.ViewModelProviders;
import android.os.Bundle;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.Toolbar;
import android.view.View;

import com.codearms.maoqiqi.app.R;
import com.codearms.maoqiqi.app.ViewModelFactory;
import com.codearms.maoqiqi.app.base.BaseActivity;
import com.codearms.maoqiqi.app.tasks.TasksActivity;

/**
 * Displays an add or edit task screen.
 * Author: fengqi.mao.march@gmail.com
 * Date: 2019/3/12 13:14
 */
public class AddEditTaskActivity extends BaseActivity implements View.OnClickListener {

    private AddEditTaskViewModel addEditTaskViewModel;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_add_edit_task);

        String taskId = getIntent().getStringExtra(TasksActivity.EXTRA_TASK_ID);

        findViewById(R.id.fabAddEditTask).setOnClickListener(this);

        Toolbar toolbar = findViewById(R.id.toolbar);
        toolbar.setTitle(taskId == null ? R.string.add_task : R.string.edit_task);
        setSupportActionBar(toolbar);
        if (getSupportActionBar() != null)
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);

        AddEditTaskFragment addEditTaskFragment = (AddEditTaskFragment) getSupportFragmentManager().findFragmentById(R.id.contentFrame);
        if (addEditTaskFragment == null) {
            addEditTaskFragment = AddEditTaskFragment.newInstance();
            getSupportFragmentManager().beginTransaction().add(R.id.contentFrame, addEditTaskFragment).commit();
        }

        ViewModelFactory factory = ViewModelFactory.getInstance(getApplication());
        addEditTaskViewModel = ViewModelProviders.of(this, factory).get(AddEditTaskViewModel.class);
        addEditTaskViewModel.setTaskId(taskId);

        addEditTaskFragment.setViewModel(addEditTaskViewModel);
    }

    @Override
    public void onClick(View v) {
        addEditTaskViewModel.addTask();
    }

    @Override
    public boolean onSupportNavigateUp() {
        onBackPressed();
        return true;
    }
}

File: AndroidArchitecture/mvvm-live-kotlin/src/main/java/com/codearms/maoqiqi/app/addedittask/AddEditTaskActivity.kt
package com.codearms.maoqiqi.app.addedittask

import android.os.Bundle
import android.view.View
import androidx.activity.viewModels
import com.codearms.maoqiqi.app.R
import com.codearms.maoqiqi.app.base.BaseActivity
import com.codearms.maoqiqi.app.tasks.TasksActivity
import com.codearms.maoqiqi.app.utils.getViewModelFactory
import com.codearms.maoqiqi.app.utils.setToolbar

/**
 * Displays an add or edit task screen.
 * Author: fengqi.mao.march@gmail.com
 * Date: 2019/3/12 13:14
 */
class AddEditTaskActivity : BaseActivity(), View.OnClickListener {

    private val addEditTaskViewModel: AddEditTaskViewModel by viewModels { getViewModelFactory() }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_add_edit_task)

        val taskId: String? = intent.getStringExtra(TasksActivity.EXTRA_TASK_ID)

        findViewById<View>(R.id.fabAddEditTask).setOnClickListener(this)

        setToolbar(R.id.toolbar) {
            setTitle(if (taskId == null) R.string.add_task else R.string.edit_task)
            setDisplayHomeAsUpEnabled(true)
        }
    }

    override fun onClick(v: View) {
        addEditTaskViewModel.addTask()
    }

    override fun onSupportNavigateUp(): Boolean {
        onBackPressed()
        return true
    }
}